File: Graphs/Detect_cycle_using_BFS.py
from collections import deque


class Graph:

    def __init__(self, V):

        self.V = V
        self.l = [[] for _ in range(V)]


    def addEdge(self, u, v):

        self.l[u].append(v)
        self.l[v].append(u)


    def DetectCycle(self, src, vis):

        q = deque()

        vis[src] = True

        q.append((src, -1))


        while q:

            node, parent = q.popleft()


            for adjNode in self.l[node]:

                if not vis[adjNode]:

                    vis[adjNode] = True

                    q.append((adjNode, node))


                elif adjNode != parent:

                    return True


        return False


    def isCycle(self):

        vis = [False] * self.V


        for i in range(self.V):

            if not vis[i]:

                if self.DetectCycle(i, vis):

                    return True


        return False



# main function

g = Graph(5)

g.addEdge(0, 1)
g.addEdge(1, 2)
g.addEdge(2, 3)
g.addEdge(3, 4)
g.addEdge(4, 1)   # cycle


if g.isCycle():

    print("Cycle exists")

else:

    print("No cycle")

File: Graphs/Detect_cycle_using_BFS.cpp
#include <iostream>
#include <vector>
#include <list>
#include <queue>
using namespace std;

class Graph
{

    int V;
    list<int> *l;

    bool DetectCycle(int src, vector<bool> &vis);

public:

    Graph(int V)
    {
        this->V = V;
        l = new list<int>[V];
    }

    void addEdge(int u, int v);

    bool isCycle();
};


void Graph::addEdge(int u, int v)
{
    l[u].push_back(v);
    l[v].push_back(u);
}


bool Graph::DetectCycle(int src, vector<bool> &vis)
{

    queue<pair<int,int>> q;

    vis[src] = true;

    q.push({src,-1});

    while(!q.empty())
    {

        int node = q.front().first;
        int parent = q.front().second;

        q.pop();

        for(int adjNode : l[node])
        {

            if(!vis[adjNode])
            {

                vis[adjNode] = true;

                q.push({adjNode,node});
            }

            else if(adjNode != parent)
            {

                return true;
            }
        }
    }

    return false;
}


bool Graph::isCycle()
{

    vector<bool> vis(V,false);

    for(int i=0;i<V;i++)
    {

        if(!vis[i])
        {

            if(DetectCycle(i,vis))
            {

                return true;
            }
        }
    }

    return false;
}



int main()
{

    Graph g(5);

    g.addEdge(0,1);
    g.addEdge(1,2);
    g.addEdge(2,3);
    g.addEdge(3,4);
    g.addEdge(4,1); // creates cycle


    if(g.isCycle())
    {

        cout<<"Cycle exists"<<endl;
    }

    else
    {

        cout<<"No cycle"<<endl;
    }

}